### PR TITLE
[Training] SG with Momentum Optimizer

### PR DESCRIPTION
**Pull request by wschin**
PR #2314 is a single place for reviewing the whole training story.

This PR is similar to #1955 but adds another signature for stochastic gradient method with momentum. It covers [Pytorch's  SGD](https://pytorch.org/docs/stable/_modules/torch/optim/sgd.html) and [Tensorflow's MomentomOptimizer](https://www.tensorflow.org/api_docs/python/tf/train/MomentumOptimizer). Note that `SGD` is never a descent algorithm so I call it `SG` in my PR's title.

The script below is used to verify if the design can cover both of TF and Pytorch cases.
```python
import itertools
import numpy as np
import torch
from torch import optim
from torch import nn
import torch.random as rnd

iteration_count = 3
n = 10 # number of features
l = 5 # number of data points (aka batch size)

X_ = torch.randn(l, n)
Y_ = torch.randn(l, 1)

# An example implementation of ONNX Momentum when "nesterov" attribute is false.
def apply_momentum(t, r, x, g, v, norm_coefficient, alpha_, beta_):  # type: ignore
    # Add gradient of regularization term.
    g_regularized = norm_coefficient * x + g
    # Coefficient of gradient should be 1 at the first iteration.
    beta_adjusted = beta_ if t > 0 else 1
    # Update momentum.
    v_new = alpha_ * v + beta_adjusted * g_regularized
    # Apply SG with momentum update rule.
    x_new = x - r * v_new
    return x_new, v_new

# An example implementation of ONNX Momentum when "nesterov" attribute is true.
def apply_nesterov(t, r, x, g, v, norm_coefficient, alpha_, beta_):  # type: ignore
    # Add gradient of regularization term.
    g_regularized = norm_coefficient * x + g
    # Coefficient of gradient should be 1 at the first iteration.
    beta_adjusted = beta_ if t > 0 else 1
    v_new = alpha_ * v + beta_adjusted * g_regularized
    # Apply Nesterov with momentum update rule.
    x_new = x - r * (g_regularized + alpha_ * v_new)
    return x_new, v_new

def show_pytorch(lr, alpha, beta, weight_decay, nesterov):
    X = X_.clone()
    Y = Y_.clone()

    rnd.manual_seed(0)
    model = nn.Sequential(
        nn.Linear(n, 1, bias=False)
    )

    loss_fn = nn.MSELoss(reduction='sum')

    solver = optim.SGD(model.parameters(), lr=lr, momentum=alpha, dampening=1-beta,
            weight_decay=weight_decay, nesterov=nesterov)

    results = []
    for t in range(iteration_count):
        Y_pred = model(X)
        loss = loss_fn(Y_pred, Y)
        results.append(float(loss))
        model.zero_grad()
        loss.backward()
        solver.step()
    return results

def show_tensorflow(lr, alpha, nesterov):
    import tensorflow as tf
    rnd.manual_seed(0)
    layer = nn.Linear(n, 1, bias=False)

    X = tf.placeholder('float', shape=[l, n])
    W = tf.Variable(torch.Tensor.numpy(layer.weight.detach()))
    Y = tf.placeholder('float', shape=[l, 1])
    Y_pred = tf.matmul(X, W, transpose_b=True)
    loss = tf.reduce_sum(tf.square(Y - Y_pred))
    optimizer = tf.train.MomentumOptimizer(learning_rate=lr, momentum=alpha, use_nesterov=nesterov)
    minimizer = optimizer.minimize(loss)

    sess = tf.Session()
    init = tf.global_variables_initializer()
    sess.run(init)

    results = []
    for t in range(iteration_count):
        loss_value = sess.run(loss, {X: X_, Y: Y_})
        results.append(float(loss_value ))
        sess.run([minimizer], {X: X_, Y: Y_})

    return results


def show_mine(lr, alpha, beta, weight_decay, nesterov):
    X = X_.clone()
    Y = Y_.clone()

    rnd.manual_seed(0)
    model = nn.Sequential(
        nn.Linear(n, 1, bias=False)
    )

    loss_fn = nn.MSELoss(reduction='sum')

    solver = optim.SGD(model.parameters(), lr=lr, momentum=alpha, dampening=1-beta,
            weight_decay=weight_decay, nesterov=nesterov)

    results = []
    for t in range(iteration_count):
        Y_pred = model(X)
        loss = loss_fn(Y_pred, Y)
        results.append(float(loss))
        model.zero_grad()
        loss.backward()

        with torch.no_grad():
            for param in model.parameters():
                if 'momentum_buffer' not in solver.state[param]:
                    solver.state[param]['momentum_buffer'] = torch.zeros_like(param)
                
                if (nesterov):
                    new_tensor, new_state = apply_nesterov(t=t,
                            r=lr, x=param.data, g=param.grad.data,
                            v=solver.state[param]['momentum_buffer'].data,
                            norm_coefficient=weight_decay,
                            alpha_=alpha, beta_=beta)
                else:
                    new_tensor, new_state = apply_momentum(t=t,
                            r=lr, x=param.data, g=param.grad.data,
                            v=solver.state[param]['momentum_buffer'].data,
                            norm_coefficient=weight_decay,
                            alpha_=alpha, beta_=beta)
                solver.state[param]['momentum_buffer'].data = new_state.data
                param.data = new_tensor.data

    return results


# Compare TF SG and its ONNX counterpart.
lr_pool = [0.1, 0.001]
alpha_pool = [1.2, 0.8, 0.01]
beta_pool = [1]
weight_decay_pool = [0]
nesterov_pool = [True, False]

for lr_, alpha_, beta_, weight_decay_, nesterov_ in itertools.product(lr_pool, alpha_pool, beta_pool,
                                                                      weight_decay_pool, nesterov_pool):
    tf_result = show_tensorflow(lr=lr_, alpha=alpha_, nesterov=nesterov_)
    mine_result = show_mine(lr=lr_, alpha=alpha_, beta=beta_, weight_decay=weight_decay_, nesterov=nesterov_)
    assert np.allclose(tf_result, mine_result)


# Compare Pytorch SG with momentum and its ONNX counterpart.
lr_pool = [0.1, 0.001]
alpha_pool = [1.2, 0.8, 0.01]
beta_pool = [1, 0.2, 0.5]
weight_decay_pool = [0, 0.01, 0.1, 1]
nesterov_pool = [False]
for lr_, alpha_, beta_, weight_decay_, nesterov_ in itertools.product(lr_pool, alpha_pool, beta_pool,
                                                                      weight_decay_pool, nesterov_pool):
    torch_result = show_pytorch(lr=lr_, alpha=alpha_, beta=beta_, weight_decay=weight_decay_, nesterov=nesterov_)
    mine_result = show_mine(lr=lr_, alpha=alpha_, beta=beta_, weight_decay=weight_decay_, nesterov=nesterov_)
    assert np.allclose(torch_result, mine_result)


# Compare Pytorch SG with Nesterov momentum and its ONNX counterpart.
lr_pool = [0.1, 0.001]
alpha_pool = [1.2, 0.8, 0.01]
beta_pool = [1]
weight_decay_pool = [0, 0.01, 0.1, 1]
nesterov_pool = [True]
for lr_, alpha_, beta_, weight_decay_, nesterov_ in itertools.product(lr_pool, alpha_pool, beta_pool,
                                                                      weight_decay_pool, nesterov_pool):
    torch_result = show_pytorch(lr=lr_, alpha=alpha_, beta=beta_, weight_decay=weight_decay_, nesterov=nesterov_)
    mine_result = show_mine(lr=lr_, alpha=alpha_, beta=beta_, weight_decay=weight_decay_, nesterov=nesterov_)
    assert np.allclose(torch_result, mine_result)
```

### CODE REVIEW

**Comment by gramalingam** [comment]
I don't think these ops are control-flow ops … so they should probably go into a more appropriate defs file, instead of the control-flow defs file.

File: onnx/defs/controlflow/defs.cc [hunk 1]
```diff
@@ -775,4 +775,94 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+        .Input(
+            4,
+            "inputs",
+            "It sequentially contains the current values of optimized tensors and then their "
+            "momentum tensors. For example, if two tensor \"X_1\" and \"X_2\" are optimized, The "
+			"expected input list would be [\"X_1\", \"X_2\", momentum of \"X_1\", momentum of \"X_2\"].",
+            "T2",
+            OpSchema::Variadic,
+            false)
+        .Output(
```
Comment (wschin) [inline onnx/defs/controlflow/defs.cc]: Need an attribute to control if Nestrove momentum is used.

File: onnx/defs/controlflow/defs.cc [hunk 2]
```diff
@@ -775,4 +775,94 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+      // Add gradient of 0.5 * Lambda * ||X||_F^2, where ||X||_F is the Frobenius norm.
+      G_regularized = Lambda * X + G;
+
+      // Compute the current momentum based on previous momentum and the current gradient.
+      V_new = Alpha * V + Beta * G;
+
+      // Update X.
+      X_new = X - R * V_new
+
+    Pseudo code for SG with Nestrove's Momentum:
```
Comment (wschin) [inline onnx/defs/controlflow/defs.cc]:
```suggestion
    Pseudo code for SG with Nesterov's Momentum:
```

**Comment by wschin** [comment]
@gramalingam, I will move this file to the right place after #1955 is merged.
